### PR TITLE
CUMULUS-1171 Make provider handling consistent across protocols

### PR DESCRIPTION
**Pull request by yjpa7145**
Addresses [CUMULUS-1171 - provider.host is a hostname in some instances, and a URL in others](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1171)

NOTE: This will be a breaking change in the `@cumulus/api` and `@cumulus/ingest` packages.

- Fixed provider handling in the API to make it consistent between protocols. Before this change, FTP providers were configured using the `host` and `port` properties. HTTP providers ignored `port` and `protocol`, and stored an entire URL in the `host` property. Updated the API to only accept valid hostnames or IP addresses in the `provider.host` field. Updated ingest code to properly build HTTP and HTTPS URLs from `provider.protocol`, `provider.host`, and `provider.port`.
- The default provider port was being set to 21, no matter what protocol was being used. Removed that default.
- Added `@cumulus/common` API documentation to `packages/common/docs/API.md`
- Added an `npm run build-docs` task to `@cumulus/common`
- Added `@cumulus/common/string#isValidHostname()`
- Added `@cumulus/common/string#match()`
- Added `@cumulus/common/string#matches()`
- Added `@cumulus/common/string#toLower()`
- Added `@cumulus/common/string#toUpper()`
- Added `@cumulus/common/URLUtils#buildURL()`
- Added `@cumulus/common/util#isNil()`
- Added `@cumulus/common/util#isNull()`
- Added `@cumulus/common/util#isUndefined()`
- Added `@cumulus/common/util#negate()`

### CODE REVIEW

**Comment by yjpa7145** [comment]
This PR looks a lot bigger than it actually is.  It's mostly new documentation and tests.

File: example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js [hunk 1]
```diff
@@ -25,37 +25,31 @@ const executionModel = new Execution();
 
 // Note: This test runs in serial due to the logs endpoint tests
 
-describe('The Discover Granules workflow with https Protocol', () => {
+// Disabled until we're acutally using https
```
Comment (laurenfrederick) [inline example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js]: Do we not have https support?

File: packages/common/string.js [hunk 2]
```diff
@@ -25,12 +45,94 @@ const unicodeEscape = (str, regex = /[\s\S]/g) => str.replace(regex, unicodeEsca
...
+ *
+ * @param {RegExp} regexp - the pattern to match against
+ * @param {string} str - the string to match against
+ * @returns {boolean} true if the pattern matches the string, false otherwise
+ *
+ * @static
+ * @kind function
+ *
+ * @example
+ * const isCapitalized = matches(/^[A-Z]/);
```
Comment (laurenfrederick) [inline packages/common/string.js]: This example does not match.

File: packages/common/string.js [hunk 3]
```diff
@@ -25,12 +45,94 @@ const unicodeEscape = (str, regex = /[\s\S]/g) => str.replace(regex, unicodeEsca
...
 
+/**
+ * Converts string, as a whole, to lower case just like String#toLowerCase
+ *
+ * @param {string} str - the string to convert
+ * @returns {string} the lower-cased string
+ *
+ * @static
+ */
+const toLower = (str) => str.toLowerCase();
```
Comment (laurenfrederick) [inline packages/common/string.js]: Do all these new functions have unit tests?

File: packages/common/util.js [hunk 4]
```diff
@@ -74,3 +87,36 @@ exports.omit = (objectIn, keys) => {
...
 };
+
+/**
+ * Creates a function that returns the opposite of the predicate function.
+ *
+ * @param {Function} predicate - the predicate to negate
+ * @returns {Function} the new negated function
+ * @kind function
+ */
+exports.negate = (predicate) => (...args) => !predicate.apply(this, args);
```
Comment (laurenfrederick) [inline packages/common/util.js]: Do these have unit tests?

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,32 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+  - Added `@cumulus/common/string#toLower()`
+  - Added `@cumulus/common/string#toUpper()`
+  - Added `@cumulus/common/URLUtils#buildURL()`
+  - Added `@cumulus/common/util#isNil()`
+  - Added `@cumulus/common/util#isNull()`
+  - Added `@cumulus/common/util#isUndefined()`
+  - Added `@cumulus/common/util#negate()`
+
+### Changed
+- CUMULUS-1171
```
Comment (laurenfrederick) [inline CHANGELOG.md]: This needs to be called out as a breaking change with steps to migrate. i.e. update all providers with all required fields. What will happen if they have an existing provider that does not have all the fields and they run a workflow? Will they get an error that explains what the issue is?

File: example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js (hunk 1 above)
Comment (yjpa7145) [reply example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js]: No, we don't have an HTTPS provider to download from in our integration tests.

File: packages/common/string.js (hunk 2 above)
Comment (yjpa7145) [reply packages/common/string.js]:
```
> const stringUtils = require('./string')
undefined
> const isCapitalized = stringUtils.matches(/^[A-Z]/);
undefined
> isCapitalized('Joe');
true
```

File: packages/common/util.js (hunk 4 above)
Comment (yjpa7145) [reply packages/common/util.js]: Added in ad2fcc9bd

File: packages/common/string.js (hunk 3 above)
Comment (yjpa7145) [reply packages/common/string.js]: Added in 3cdaef3d1

File: CHANGELOG.md (hunk 5 above)
Comment (yjpa7145) [reply CHANGELOG.md]: As of 03f9fcd19, trying to discover or download using an invalid `provider.host` will throw a `TypeError` with a descriptive message.

File: packages/common/string.js (hunk 2 above)
Comment (laurenfrederick) [reply packages/common/string.js]: oh sorry i read this wrong. I found the function a bit confusing.

**Pull request merged**